### ISSUE TITLE
Guidance on Replicating Paper Results and Using `--search-algorithm human`

### ISSUE THREAD

**Issue by MTDickens** [belalugaX/rigid-origami#3]
Hello,

First, thank you for this very helpful tool! I'm a researcher from ZJU, currently collaborating with a postdoc and several undergrads from MIT, NJU and our school on a project to create a (hopefully) large-scale origami dataset. Your work is a fantastic starting point for us.

We are trying to replicate some of the results from your accompanying paper to ensure we understand the tool's workflow before generating our own data. We're running into difficulties with setting the correct initial crease patterns for specific examples.

We have two main questions:

**1. Replicating Paper Results (e.g., Human Face, Chair Fold)**

The paper's Appendix C mentions, "The human face is initialized with a single crease line just above the tip of the nose." Similarly, Figure 8 shows the seeding pattern for the chair fold.

We assume that replicating these results requires setting specific initial parameters like `base`, `start-sequence`, and `seed pattern size`. However, we're unsure how to translate the descriptions from the paper into the correct command-line arguments.

*   Could you please provide an example command or the specific `--start-sequence` value needed to replicate the **human face** result?
*   Similarly, what would be the initial setup for the **chair fold** example?

**2. Using the `--search-algorithm human`**

For creating more complex shapes, we anticipate needing to define the initial base or seed pattern manually. We noticed the `--search-algorithm human` option, which seems perfect for this "human-in-the-loop" approach.

*   Could you explain the workflow for using this mode? Like how we are supposed to input the numbers? And what do the numbers/matrices mean? A simple usage example would be incredibly helpful.

Any guidance on these points would be greatly appreciated. Our goal is to fully understand the tool so we can effectively use it to build our dataset.

Thank you for your time and for developing this great project!

**Comment by MTDickens** [comment]
Hello again,

Following up on my GitHub issue, we had a few broader, more strategic questions about our project that we were hoping to get your perspective on.

Please feel absolutely no obligation to answer these—we know they go beyond typical user support and that your time is valuable. However, as the creator of this tool, your insights would be incredibly valuable to us as we scope out our work.

1.  **On Target Object Difficulty:** From your experience, are certain categories of objects (e.g., polyhedra, furniture, household goods, animals) significantly easier to generate crease patterns for than others? We are trying to decide on a good starting point for our team and were wondering if there's a "low-hanging fruit" category you might recommend.

2.  **On Scalable Dataset Creation Strategy:** We are currently debating two potential paths for scaling up our dataset generation, and we would be grateful for your opinion on their feasibility:
    *   **Shape Approximation:** Using existing 3D models (`.obj` files) and optimizing against a metric like the Hausdorff distance.
    *   **"Shape Imagination":** Manually defining custom reward functions for each target shape, perhaps augmented by modern AI models (e.g., Gemini 2.5 Pro) to help define or interpret rewards, in case it's too time-consuming.
    Does one of these approaches seem more promising or practical to you for creating a dataset of a few hundred/thousand examples?

3.  **On Project Feasibility and Effort:** Finally, we noticed that there are only 8 shapes presented in your paper, and also the need for carefully tuned constraints for each one. This leads to our main concern about feasibility: do you believe it's practical for a small team (~6-7 undergrads with very limited research experience) to generate a large-scale dataset in a reasonable timeframe? We're not asking for exact hours, but any general insight you could share on the effort involved in the constraint-tuning process for your paper's examples would be immensely helpful for us in setting realistic goals.

Again, please don't feel pressured to provide a detailed response. Any brief thoughts or advice you might have would be more than enough and greatly appreciated.

Thank you once again for your time and for creating such a powerful tool.

**Comment by belalugaX** [comment]
Hi Max,

Thanks for reaching out. 

Let me provide short answers on your points. I will come back later with example commands etc. (need some days to catch up on the issue, I am traveling these days).

1. A general rule: the design and problem space naturally explodes in complexity with increased board size. It is therefore essential to constrain the space properly according to a particular target shape. Shape primitives, furniture or other simple objects seem to be quite easier than complex shapes (like the human face obviously), because of a) easier topologies require smaller boards, b) use of symmerty constraints significantly decrease the search space. (accurate) Shape approximation of complex topologies is very challenging, depending on the reward function, because you most likely will need larger board sizes -> a more refined mesh in order to properly approximate a given target shape. 

Reflecting on our paper’s journey, a core idea was to descretize the design space and come up with what we called the “origami game.” This allowed us to reach acceptable results with limited computing resources. This however, does already quite constrain the design space upfront. Hope that makes sense… :)

2. I would definitely recommend you dedicate a limited time budget to experiment with both shape approximation and imagination to get a basic intuition. Shape imagination is naturally more forgiving than shape approximation, because with imagination you leave space for the agents to “surprise" you with some creative solutions that may also just serve as  inspiration for enhanced origami foldable design of whatever objects like furniture.

3. Regarding “large dataset” generation: How many shapes do you have in mind? 

It does definitely make sense to replicate the results from our paper to get a solid understanding of the current limitations of our approach. The "human in the loop” option is exactly meant for that – playing the origami game manually. That’s how it basically works (recalling from my memory):

(0. you may first need to place the seed pattern, depending on the config used)

1. select a given extendable # vertex from the list of nodes available -> enter
(two options per vertex due to their rigid-body-mode – up-/downward folding. That’s the twofold solutios to the PTU equations)
2. select three extension vertices from the given list. 
2.1 select first # vertex -> enter
2.2 select second # vertex -> enter
2.3 select third # vertex -> enter
3. see the current board config. Symmerty rules if applicable will be executed after each full extension (steps 1 through 3)

Hope that helps you a bit further now and hope you get it running.

I will come back to you in a few days.

Best,

Jeremia

**Comment by MTDickens** [comment]
Thank you! Looking forward to your tutorials.

Also, regarding the number of origamis, after a brief discussion with my coworker (who has much more model training experience than me), we decide that we can make 250 for training and another 250 for testing. I would like to know if that sound feasible to you, according to your experience with data collection.

**Comment by belalugaX** [comment]
Hi Max,

Many thanks for your patience.

I did run the code on our various search algos on my local machine (Mac M3): RDM, DFTS, BFTS, MCTS, PPO. The code should run off-the-shelf, but most likely you need to adjust depending on your system architecture. For my Mac M3 I also had to reinstall/adjust some dependencies for arm64. Please let me know if you face any issues here.


Regarding your latest question: It is easy to generate patterns, but difficult to generate good ones and hard to come up with proper solutions for complex shapes beyond primitives. Not sure what exactly you have in mind. But I suggest you try the following:

1. try the human in the loop example below

2. run the example commands on the given example shapes (see attached files)

3. play around with the arguments and other/your own shape primitives

4. Then increase the board size "—board-length" (edge length of the board) to get better approximations of the human face. You will need to scale the human face shape and play around with the “—base" argument and positioning, etc. You will quickly get a solid idea of the problem's complexity. If you manage to handle the scaling up challenge somehow efficiently, by means of some pretrained models, cascading or break the problem down into patches etc., I expect you can achieve pretty interesting stuff.



Regarding the “human in the loop” example. Please try to run this command here: 

python main.py --objective=shape-approx --search-algorithm=human --num-steps=500000 --board-length=9 --num-symmetries=2 --optimize-psi --target="default_targets/small_cube.obj" --allow-source-action --base=quad 

You will get the output (sorry that’s not very intuitive): first matrix BOARD CURRENT STATE, second matrix AVAILABLE ACTIONS, third vector ACTION VECTOR. 

Type in an action, like “43” and enter. Follow this list of actions here to get the example pattern below: [ 43, 13, 17, 12,13,9, 12, 8, 17, 12, 21, 16, 9, 8, 4, 23] -> “23” (i.e. vertex “0”) is the terminal action here.

Results will be saved to the “results” folder. You get different objs with various states of the folding motion.

But why does the action vector contain a “43” that is not listed in the action space matrix? Because the first action “select an extendable vertex”, where every extendable vertex can be folded up (blue)/downwards (olive green) – i.e. the rigid body mode from our paper. 18 and 43 mean the same vertex but in different rigid body mode. 


Example commands from our paper: please see the attached file.

Please give me an update if it all worked out or in case of any further questions. 

Best,

Jeremia

￼
